app/src/main/java/com/example/android/architecture/blueprints/todoapp/di/ApplicationComponent.kt

```diff
@@ -0,0 +1,84 @@
+package com.example.android.architecture.blueprints.todoapp.di
+
+import android.content.Context
+import androidx.room.Room
+import com.example.android.architecture.blueprints.todoapp.TodoApplication
+import com.example.android.architecture.blueprints.todoapp.data.source.DefaultTasksRepository
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository
+import com.example.android.architecture.blueprints.todoapp.data.source.local.TasksLocalDataSource
+import com.example.android.architecture.blueprints.todoapp.data.source.local.ToDoDatabase
+import com.example.android.architecture.blueprints.todoapp.data.source.remote.TasksRemoteDataSource
+import dagger.BindsInstance
+import dagger.Component
+import dagger.Module
+import dagger.Provides
+import dagger.android.AndroidInjector
+import dagger.android.support.AndroidSupportInjectionModule
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+@Component(
+    modules = [
+        ApplicationModule::class,
+        AndroidSupportInjectionModule::class,
+        TasksModule::class,
+        TaskDetailModule::class,
+        AddEditTaskModule::class,
+        StatisticsModule::class
+    ])
+interface ApplicationComponent : AndroidInjector<TodoApplication> {
+
+    @Component.Factory
+    interface Factory {
+        fun create(@BindsInstance applicationContext: Context): ApplicationComponent
+    }
+
+    val tasksRepository: TasksRepository
+}
+
+@Module
+class ApplicationModule {
+
+    @Singleton
+    @Provides
+    fun provideRepository(
+        @Named("TasksRemoteDataSource") tasksRemoteDataSource: TasksDataSource,
+        @Named("TasksLocalDataSource") tasksLocalDataSource: TasksDataSource,
+        ioDispatcher: CoroutineDispatcher
+    ): TasksRepository {
+        return DefaultTasksRepository(tasksRemoteDataSource, tasksLocalDataSource, ioDispatcher)
+    }
+
+    @Singleton
+    @Named("TasksRemoteDataSource")
+    @Provides
```

Comment:
1Seems only `TasksRepository` is injected in constructors. The different `TasksDataSource` and `ToDoDatabase` don't need to be annotated with `@Provides`

Reply:
I get `TasksDataSource cannot be provided without an @Provides-annotated method`